linkerd/metrics/src/counter.rs

```diff
@@ -15,75 +15,72 @@ use super::prom::{FmtLabels, FmtMetric};
 /// [`rate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#rate()
 /// [`irate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#irate()
 /// [`resets()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#resets
-#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
-pub struct Counter(u64);
-
-/// Largest `u64` that can fit without loss of precision in `f64` (2^53).
-pub(crate) const MAX_PRECISE_COUNTER: u64 = 0x20_0000_0000_0000;
+#[derive(Debug, Default)]
+pub struct Counter(AtomicU64);
 
 // ===== impl Counter =====
 
 impl Counter {
-    /// Increment the counter by one.
-    ///
-    /// This function wraps on 52-bit overflows.
-    pub fn incr(&mut self) {
-        *self += 1;
+    pub fn incr(&self) {
+        self.add(1)
+    }
+
+    pub fn add(&self, n: u64) {
+        self.0.fetch_add(n, Ordering::Release);
     }
 
-    /// Return current counter value.
+    /// Return current counter value, wrapped to be safe for use with Prometheus.
     pub fn value(&self) -> u64 {
         self.0
+            .load(Ordering::Acquire)
+            .wrapping_rem(MAX_PRECISE_VALUE + 1)
```

Comment:
oh, just wrapping here is much neater than doing a load-wrap-CAS loop as i suggested, 👍